### PR TITLE
ipam pool leak fix

### PR DESCRIPTION
**Pull request by aegal**
**Reason for Change**:
This is to address the potential for IPAM to leak IP's and have the pool be stuck in a 'in use ' state.

**Issue Fixed**:
https://github.com/Azure/azure-container-networking/issues/633

This also reintroduces some tests that were last in previous commits. 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests

**Notes**:

### CODE REVIEW

File: ipam/api.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ var (
 	errInvalidConfiguration    = fmt.Errorf("Invalid configuration")
 	errAddressPoolExists       = fmt.Errorf("Address pool already exists")
 	errAddressPoolNotFound     = fmt.Errorf("Address pool not found")
+	ErrAddressPoolNotFound     = fmt.Errorf("Address pool not found")
```
Comment (aegal) [inline ipam/api.go:19]: Remove this

File: ipam/pool.go [hunk 2]
```diff
@@ -278,7 +283,7 @@ func (as *addressSpace) newAddressPool(ifName string, priority int, subnet *net.
 func (as *addressSpace) getAddressPool(poolId string) (*addressPool, error) {
 	ap := as.Pools[poolId]
 	if ap == nil {
-		return nil, errInvalidPoolId
+		return nil, ErrAddressPoolNotFound
```
Comment (aegal) [inline ipam/pool.go:286]: Revert

File: ipam/pool_test.go [hunk 3]
```diff
@@ -509,7 +510,7 @@ var (
 						Pools: map[string]*addressPool{},
 					}
 					pool, err := as.getAddressPool("10.0.0.0/16")
-					Expect(err).To(Equal(errInvalidPoolId))
+					Expect(err).To(Equal(errAddressPoolNotFound))
```
Comment (aegal) [inline ipam/pool_test.go:513]: Revert

File: cni/netconfig.go [hunk 4]
```diff
@@ -69,6 +69,7 @@ type NetworkConfig struct {
 		Subnet        string `json:"subnet,omitempty"`
 		Address       string `json:"ipAddress,omitempty"`
 		QueryInterval string `json:"queryInterval,omitempty"`
+		EndpointID    string
```
Comment (aegal) [inline cni/netconfig.go:72]: Could also use container id already passed in the CNI args

Comment (aegal) [reply cni/netconfig.go:72]: Use ContainerId

File: ipam/manager_old_test.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	// Remove addr12 and add addr13 in subnet1.
+	ap, err := localAs.newAddressPool(anyInterface, anyPriority, &subnet1)
+	ap.newAddressRecord(&addr11)
+	ap.newAddressRecord(&addr13)
+
+	// Remove subnet2.
+	// Add subnet3 with addr31.
+	ap, err = localAs.newAddressPool(anyInterface, anyPriority, &subnet3)
+	ap.newAddressRecord(&addr31)
+
```
Comment (aegal) [inline ipam/manager_old_test.go:64]: Write this back into the current file.

File: ipam/pool.go [hunk 6]
```diff
@@ -484,7 +507,7 @@ func (ap *addressPool) requestAddress(address string, options map[string]string)
 	// If no address was found, return any available address.
 	if ar == nil {
 		for _, ar = range ap.Addresses {
-			if !ar.InUse && ar.ID == "" {
```
Comment (aegal) [inline ipam/pool.go:487]: Revert ID

File: ipam/pool.go [hunk 7]
```diff
@@ -546,7 +570,7 @@ func (ap *addressPool) releaseAddress(address string, options map[string]string)
 		return nil
 	}
 
-	if !ar.InUse {
+	if !ar.InUse && ar.ID == "" {
```
Comment (aegal) [inline ipam/pool.go:573]: Double check why adding this

File: ipam/manager_old_test.go (hunk 5 above)
Comment (aegal) [reply ipam/manager_old_test.go]: Reverted

File: cni/network/network.go [hunk 8]
```diff
@@ -837,6 +837,13 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
 
 	// Query the network.
 	if nwInfo, err = plugin.nm.GetNetworkInfo(networkId); err != nil {
+
+		// attempt to release address associated with this Endpoint id
+		// This is to ensure clean up is done even in failure cases
+		if err = plugin.DelegateDel(nwCfg.Ipam.Type, nwCfg); err != nil {
+			log.Printf("Network not found, attempted to release address with error:  %v", err)
```
[the remark this answers is not in the document]
Comment (aegal) [reply cni/network/network.go:844]: As discussed offline, leaving it as is

File: ipam/pool.go [hunk 9]
```diff
@@ -360,29 +371,31 @@ func (as *addressSpace) requestPool(poolId string, subPoolId string, options map
...
 	}
 
 	if err != nil {
 		log.Printf("[ipam] Failed to release pool, err:%v.", err)
 		return err
 	}
 
-	ap.RefCount--
+	if !addressesInUse {
+		ap.RefCount--
```
[the remark this answers is not in the document]
Comment (aegal) [reply ipam/pool.go:392]: I also thought of that as well. That was my only worry, if a process is using the pool but hasn't allocated an address yet and by setting this to zero, this pool is deleted before that happens. Not sure what exact scenario that may occur but it seems it was designed with that in mind.

**Comment by aegal** [comment]
> I didnt see in releaseaddress id is reset to empty? how is it handled?

Discussed offline, it is already handled in the releaseaddress

File: cni/ipam/ipam_test.go [hunk 10]
```diff
@@ -203,6 +277,54 @@ var (
...
+					arg.StdinData = getStdinData("0.4.0", "", "")
+					err = plugin.Add(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+			})
 		})
 	})
 )
+
+func GetFirstAddress() string {
```
Comment (tamilmani1989) [inline cni/ipam/ipam_test.go:292]: why this is named as GetFirstAddress

File: ipam/pool.go [hunk 11]
```diff
@@ -159,6 +159,9 @@ func (am *addressManager) setAddressSpace(as *addressSpace) error {
 	if !ok {
 		am.AddrSpaces[as.Id] = as
 	} else {
+		// merges the address set refreshed from the source
+		// and the ones we have currently in this address space
+		log.Printf("[ipam] merging address space")
```
Comment (tamilmani1989) [inline ipam/pool.go:164]: can you as add log before am.save() saying "saving ipam state"

File: cni/ipam/ipam_test.go (hunk 10 above)
Comment (aegal) [reply cni/ipam/ipam_test.go:292]: Yeah, I guess calling it GetNextAddress would be more appropriate

File: ipam/pool.go (hunk 11 above)
Comment (aegal) [reply ipam/pool.go:164]: Done

**Comment by EdmondAndy** [comment]
Hi,

see the change for "ipam pool leak fix" has been merged for CNI issues.
Can you please confirm which aks version will include this fix? thanks

 Ed

**Pull request merged**